app/src/main/java/pm/gnosis/heimdall/ui/safe/create/ConfirmSafeRecoveryPhraseViewModel.kt

```diff
@@ -0,0 +1,96 @@
+package pm.gnosis.heimdall.ui.safe.create
+
+import io.reactivex.Completable
+import io.reactivex.Single
+import io.reactivex.functions.BiFunction
+import io.reactivex.schedulers.Schedulers
+import pm.gnosis.heimdall.data.remote.RelayServiceApi
+import pm.gnosis.heimdall.data.remote.models.RelaySafeCreationParams
+import pm.gnosis.heimdall.data.repositories.GnosisSafeRepository
+import pm.gnosis.mnemonic.Bip39
+import pm.gnosis.model.Solidity
+import pm.gnosis.models.Transaction
+import pm.gnosis.svalinn.accounts.base.repositories.AccountsRepository
+import pm.gnosis.svalinn.accounts.utils.hash
+import pm.gnosis.svalinn.common.utils.Result
+import pm.gnosis.svalinn.common.utils.mapToResult
+import pm.gnosis.svalinn.security.EncryptionManager
+import pm.gnosis.utils.asBigInteger
+import pm.gnosis.utils.words
+import java.math.BigInteger
+import java.nio.charset.Charset
+import java.security.SecureRandom
+import javax.inject.Inject
+
+class ConfirmSafeRecoveryPhraseViewModel @Inject constructor(
+    private val accountsRepository: AccountsRepository,
+    private val bip39: Bip39,
+    private val encryptionManager: EncryptionManager,
+    private val relayServiceApi: RelayServiceApi,
+    private val gnosisSafeRepository: GnosisSafeRepository
+) : ConfirmSafeRecoveryPhraseContract() {
+    private lateinit var mnemonic: String
+
+    private lateinit var chromeExtensionAddress: Solidity.Address
+
+    private val secureRandom = SecureRandom()
+
+    override fun setup(encryptedMnemonic: String, chromeExtensionAddress: Solidity.Address): Single<List<String>> =
+        Completable.fromAction {
+            val mnemonic = encryptionManager.decrypt(EncryptionManager.CryptoData.fromString(encryptedMnemonic)).toString(Charset.forName("UTF-8"))
+            if (mnemonic.words().size != 12) throw IllegalStateException("Invalid mnemonic")
+            this.mnemonic = mnemonic
+            this.chromeExtensionAddress = chromeExtensionAddress
+        }
+            .andThen(loadShuffledWords())
+            .subscribeOn(Schedulers.io())
+
+    private fun loadShuffledWords(): Single<List<String>> =
+        Single.fromCallable {
+            mnemonic.words().shuffled()
+        }.subscribeOn(Schedulers.computation())
+
+    override fun isCorrectSequence(words: List<String>): Single<Result<Boolean>> =
+        Single.fromCallable {
+            true
+            //words.joinToString(" ") == mnemonic
+        }.subscribeOn(Schedulers.computation()).mapToResult()
+
+    override fun createSafe(): Single<Result<BigInteger>> =
+        loadSafeOwners()
+            .map { RelaySafeCreationParams(owners = it, threshold = 2, s = BigInteger(252, secureRandom)) }
+            .flatMap { relayServiceApi.safeCreation(it) }
+            .map {
+                it to Transaction(
```

Comment:
Can we check the signature / transaction returned by the service:
- Signature.s should be our generated number
- Get account for the generated signature and check the safe address
- Check that data starts with the Gnosis proxy bytecode

Reply:
Will sync with the backend in the future on the version used for the contract (for the bytecode)